clouddrift/pairs.py

```diff
@@ -0,0 +1,407 @@
+    >>> time1 = unpack(ds["time"], ds["rowsize"], rows=0).pop()
+    >>> lon2 = unpack(ds["longitude"], ds["rowsize"], rows=1).pop()
+    >>> lat2 = unpack(ds["latitude"], ds["rowsize"], rows=1).pop()
+    >>> time2 = unpack(ds["time"], ds["rowsize"], rows=1).pop()
+    >>> chance_pair(lon1, lat1, lon2, lat2, time1, time2, 6000, np.timedelta64(0))
+    (array([-87.05255 , -87.04974 , -87.04612 , -87.04155 , -87.03718 ,
+            -87.033165], dtype=float32),
+     array([28.23184 , 28.227087, 28.222498, 28.217508, 28.2118  , 28.205938],
+           dtype=float32),
+     array([-87.111305, -87.10078 , -87.090675, -87.08157 , -87.07363 ,
+            -87.067375], dtype=float32),
+     array([28.217918, 28.20882 , 28.198603, 28.187078, 28.174644, 28.161713],
+           dtype=float32),
+     array(['2012-07-21T21:30:00.524160000', '2012-07-21T22:15:00.532800000',
+            '2012-07-21T23:00:00.541440000', '2012-07-21T23:45:00.550080000',
+            '2012-07-22T00:30:00.558720000', '2012-07-22T01:15:00.567360000'],
+           dtype='datetime64[ns]'),
+     array(['2012-07-21T21:30:00.524160000', '2012-07-21T22:15:00.532800000',
+            '2012-07-21T23:00:00.541440000', '2012-07-21T23:45:00.550080000',
+            '2012-07-22T00:30:00.558720000', '2012-07-22T01:15:00.567360000'],
+           dtype='datetime64[ns]'))
+    """
+    time_present = time1 is not None and time2 is not None
+
+    # If time is provided, subset the trajectories to the overlapping times.
+    if time_present:
+        overlap1, overlap2 = pair_time_overlap(time1, time2, time_tolerance)
+        lon1 = lon1[overlap1]
+        lat1 = lat1[overlap1]
+        time1 = time1[overlap1]
+        lon2 = lon2[overlap2]
+        lat2 = lat2[overlap2]
+        time2 = time2[overlap2]
+
+    space_tolerance_degrees = np.degrees(space_tolerance / sphere.EARTH_RADIUS_METERS)
+
+    bbox_overlap1, bbox_overlap2 = pair_bounding_box_overlap(
+        lon1, lat1, lon2, lat2, space_tolerance_degrees
+    )
+
+    lon1 = lon1[bbox_overlap1]
+    lat1 = lat1[bbox_overlap1]
+    lon2 = lon2[bbox_overlap2]
+    lat2 = lat2[bbox_overlap2]
+    time1 = time1[bbox_overlap1] if time_present else None
+    time2 = time2[bbox_overlap2] if time_present else None
+
+    space_distance = pair_space_distance(lon1, lat1, lon2, lat2)
+    chance_mask = space_distance <= space_tolerance
+
+    if time_present:
+        time_distance = pair_time_distance(time1, time2)
+        chance_mask_time = time_distance <= time_tolerance
+        chance_mask = chance_mask & chance_mask_time
+
+    indices2, indices1 = np.where(chance_mask)
+
+    if time_present:
+        return (
+            lon1[indices1],
+            lat1[indices1],
+            lon2[indices2],
+            lat2[indices2],
+            time1[indices1],
+            time2[indices2],
+        )
+    else:
+        return lon1[indices1], lat1[indices1], lon2[indices2], lat2[indices2]
+
+
+def chance_pairs(
```

Comment:
chance_pairs vs chance_pair above. Could this lead to confusion?

Reply:
It could. One letter difference is too little. This is just prototype naming. Let's discuss better names.

Reply:
How about `chance_pairs_ragged`?

Reply:
it actually should be `chance_pairs_from_ragged`?